// cspell-trie-lib/dist/lib/convertToTrieRefNodes.js
import { genSequence } from 'gensequence';
import { FLAG_WORD } from './TrieNode/TrieNode.js';
const MinReferenceCount = 3;
/**
 * An iterator that will emit TrieRefNodes mostly in descending frequency
 * @param root Root of the Trie -- a DAWG is preferred to keep the number of duplicates down.
 */
export function convertToTrieRefNodes(root) {
    const eow = { f: FLAG_WORD, c: undefined };
    const tallies = new Map([[eow, 0]]);
    let count = 0;
    const cached = new Map();
    const rollupTally = new Map();
    function tally(n) {
        if (n.f && !n.c) {
            tallies.set(eow, (tallies.get(eow) || 0) + 1);
            return;
        }
        const t = tallies.get(n);
        if (t) {
            tallies.set(n, t + 1);
            return;
        }
        tallies.set(n, 1);
        for (const c of (n.c && Object.values(n.c)) || []) {
            tally(c);
        }
    }
    function rollup(n) {
        const c = rollupTally.get(n);
        if (c) {
            return c;
        }
        if (!n.c) {
            const sum = tallies.get(eow) || 0;
            rollupTally.set(n, sum);
            return sum;
        }
        const sum = Object.values(n.c).reduce((acc, v) => acc + rollup(v), tallies.get(n) || 0);
        rollupTally.set(n, sum);
        return sum;
    }
    function* walkByTallies(tallies) {
        const nodes = genSequence(tallies).filter((a) => a[1] >= MinReferenceCount);
        for (const [n] of [...nodes].sort((a, b) => b[1] - a[1])) {
            yield* walkByRollup(n);
        }
    }
    function* walkByRollup(n) {
        if (cached.has(n))
            return;
        if (n.f && !n.c) {
            cached.set(n, cached.get(eow));
            return;
        }
        const children = ((n.c && Object.values(n.c)) || []).sort((a, b) => (rollupTally.get(b) || 0) - (rollupTally.get(a) || 0));
        for (const c of children) {
            yield* walkByRollup(c);
        }
        cached.set(n, count++);
        yield convert(n);
    }
    function convert(n) {
        const { f, c } = n;
        const r = c
            ? Object.entries(c)
                .sort((a, b) => (a[0] < b[0] ? -1 : 1))
                .map(([s, n]) => [s, cached.get(n)])
            : undefined;
        const rn = r ? (f ? { f, r } : { r }) : { f };
        return rn;
    }
    function* walk(root) {
        cached.set(eow, count++);
        yield convert(eow);
        // Emit the highest referenced nodes first.
        yield* walkByTallies(tallies);
        // Emit the rest.
        yield* walkByRollup(root);
    }
    tally(root);
    rollup(root);
    return walk(root);
}
