// cspell-trie-lib/dist/lib/TrieNode/TrieNodeBuilder.js
import { insertWordsAtCursor } from '../Builder/index.js';
import { defaultTrieInfo } from '../constants.js';
import { assert } from '../utils/assert.js';
import { mergeOptionalWithDefaults } from '../utils/mergeOptionalWithDefaults.js';
import { TrieNodeTrie } from './TrieNodeTrie.js';
const EOW = Object.freeze({ f: 1, k: true });
export class TrieNodeBuilder {
    _cursor;
    root = { ...defaultTrieInfo, c: Object.create(null) };
    wordToCharacters = (word) => [...word];
    setOptions(options) {
        const opts = mergeOptionalWithDefaults(options, this.root);
        Object.assign(this.root, opts);
        return opts;
    }
    build() {
        return new TrieNodeTrie(this.root);
    }
    getCursor() {
        this._cursor ??= this.createCursor();
        return this._cursor;
    }
    createCursor() {
        const nodes = [this.root, EOW];
        const eow = EOW;
        assert(Object.keys(this.root.c).length === 0, 'The Trie MUST be empty for cursors to work.');
        const stack = [{ n: this.root, c: '' }];
        let currNode = this.root;
        let depth = 0;
        const insertChar = (char) => {
            // console.warn('i %o', char);
            if (currNode.k) {
                const s = stack[depth];
                const { k: _, c, ...copy } = currNode;
                currNode = s.n.c[s.c] = copy;
                if (c) {
                    currNode.c = Object.assign(Object.create(null), c);
                }
                nodes.push(currNode);
            }
            const c = currNode.c || Object.create(null);
            currNode.c = c;
            const n = currNode;
            const next = (c[char] = c[char] || {});
            nodes.push(next);
            ++depth;
            const s = stack[depth];
            if (s) {
                s.n = n;
                s.c = char;
            }
            else {
                stack.push({ n, c: char });
            }
            currNode = next;
        };
        const markEOW = () => {
            // console.warn('$');
            if (!currNode.c) {
                // no children, set the parent to point to the common EOW.
                const s = stack[depth];
                s.n.c[s.c] = eow;
                if (nodes[nodes.length - 1] === currNode) {
                    nodes.pop();
                }
                currNode = eow;
            }
            else {
                currNode.f = 1;
            }
        };
        const reference = (nodeId) => {
            const s = stack[depth];
            s.n.c[s.c] = nodes[nodeId];
            nodes.pop();
        };
        const backStep = (num) => {
            if (!num)
                return;
            // console.warn('<< %o', num);
            assert(num <= depth && num > 0);
            depth -= num;
            currNode = stack[depth + 1].n;
        };
        const c = {
            insertChar,
            markEOW,
            reference,
            backStep,
        };
        return c;
    }
}
export function buildTrieNodeTrieFromWords(words) {
    const builder = new TrieNodeBuilder();
    insertWordsAtCursor(builder.getCursor(), words);
    return builder.build();
}
