// cspell-io/dist/esm/CSpellIOWeb.js
import { compareStats } from './common/stat.js';
import { ErrorNotImplemented } from './errors/errors.js';
export class CSpellIOWeb {
    readFile(_uriOrFilename) {
        throw new ErrorNotImplemented('readFile');
    }
    readFileSync(_uriOrFilename) {
        throw new ErrorNotImplemented('readFileSync');
    }
    readDirectory(_urlOrFilename) {
        throw new ErrorNotImplemented('readDirectory');
    }
    writeFile(_uriOrFilename, _content) {
        throw new ErrorNotImplemented('writeFile');
    }
    getStat(_uriOrFilename) {
        throw new ErrorNotImplemented('getStat');
    }
    getStatSync(_uriOrFilename) {
        throw new ErrorNotImplemented('getStatSync');
    }
    compareStats(left, right) {
        return compareStats(left, right);
    }
    toURL(_uriOrFilename) {
        throw new ErrorNotImplemented('toURL');
    }
    toFileURL(_urlOrFilename, _relativeTo) {
        throw new ErrorNotImplemented('toFileURL');
    }
    urlBasename(_uriOrFilename) {
        throw new ErrorNotImplemented('uriBasename');
    }
    urlDirname(_uriOrFilename) {
        throw new ErrorNotImplemented('uriDirname');
    }
}
