// cspell-io/dist/esm/VirtualFS/WrappedProviderFs.js
import { createTextFileResource, urlOrReferenceToUrl } from '../common/index.js';
import { FileType } from '../models/index.js';
import { FSCapabilityFlags, } from '../VFileSystem.js';
export function cspellIOToFsProvider(cspellIO) {
    const capabilities = FSCapabilityFlags.Stat | FSCapabilityFlags.ReadWrite | FSCapabilityFlags.ReadDir;
    const capabilitiesHttp = capabilities & ~FSCapabilityFlags.Write & ~FSCapabilityFlags.ReadDir;
    const capMap = {
        'file:': capabilities,
        'http:': capabilitiesHttp,
        'https:': capabilitiesHttp,
    };
    const name = 'CSpellIO';
    const supportedProtocols = new Set(['file:', 'http:', 'https:']);
    const fs = {
        providerInfo: { name },
        stat: (url) => cspellIO.getStat(url),
        readFile: (url) => cspellIO.readFile(url),
        readDirectory: (url) => cspellIO.readDirectory(url),
        writeFile: (file) => cspellIO.writeFile(file.url, file.content),
        dispose: () => undefined,
        capabilities,
        getCapabilities(url) {
            return fsCapabilities(capMap[url.protocol] || FSCapabilityFlags.None);
        },
    };
    return {
        name,
        getFileSystem: (url, _next) => {
            return supportedProtocols.has(url.protocol) ? fs : undefined;
        },
    };
}
function wrapError(e) {
    if (e instanceof VFSError)
        return e;
    // return new VFSError(e instanceof Error ? e.message : String(e), { cause: e });
    return e;
}
export class VFSError extends Error {
    constructor(message, options) {
        super(message, options);
    }
}
export class VFSErrorUnsupportedRequest extends VFSError {
    request;
    parameters;
    url;
    constructor(request, url, parameters) {
        super(`Unsupported request: ${request}`);
        this.request = request;
        this.parameters = parameters;
        this.url = url?.toString();
    }
}
class CFsCapabilities {
    flags;
    constructor(flags) {
        this.flags = flags;
    }
    get readFile() {
        return !!(this.flags & FSCapabilityFlags.Read);
    }
    get writeFile() {
        return !!(this.flags & FSCapabilityFlags.Write);
    }
    get readDirectory() {
        return !!(this.flags & FSCapabilityFlags.ReadDir);
    }
    get writeDirectory() {
        return !!(this.flags & FSCapabilityFlags.WriteDir);
    }
    get stat() {
        return !!(this.flags & FSCapabilityFlags.Stat);
    }
}
export function fsCapabilities(flags) {
    return new CFsCapabilities(flags);
}
export class WrappedProviderFs {
    fs;
    eventLogger;
    hasProvider;
    capabilities;
    providerInfo;
    _capabilities;
    constructor(fs, eventLogger) {
        this.fs = fs;
        this.eventLogger = eventLogger;
        this.hasProvider = !!fs;
        this.capabilities = fs?.capabilities || FSCapabilityFlags.None;
        this._capabilities = fsCapabilities(this.capabilities);
        this.providerInfo = fs?.providerInfo || { name: 'unknown' };
    }
    logEvent(method, event, traceID, url, message) {
        this.eventLogger({ method, event, url, traceID, ts: performance.now(), message });
    }
    getCapabilities(url) {
        if (this.fs?.getCapabilities)
            return this.fs.getCapabilities(url);
        return this._capabilities;
    }
    async stat(urlRef) {
        const traceID = performance.now();
        const url = urlOrReferenceToUrl(urlRef);
        this.logEvent('stat', 'start', traceID, url);
        try {
            checkCapabilityOrThrow(this.fs, this.capabilities, FSCapabilityFlags.Stat, 'stat', url);
            return new CVfsStat(await this.fs.stat(urlRef));
        }
        catch (e) {
            this.logEvent('stat', 'error', traceID, url, e instanceof Error ? e.message : '');
            throw wrapError(e);
        }
        finally {
            this.logEvent('stat', 'end', traceID, url);
        }
    }
    async readFile(urlRef, encoding) {
        const traceID = performance.now();
        const url = urlOrReferenceToUrl(urlRef);
        this.logEvent('readFile', 'start', traceID, url);
        try {
            checkCapabilityOrThrow(this.fs, this.capabilities, FSCapabilityFlags.Read, 'readFile', url);
            return createTextFileResource(await this.fs.readFile(urlRef), encoding);
        }
        catch (e) {
            this.logEvent('readFile', 'error', traceID, url, e instanceof Error ? e.message : '');
            throw wrapError(e);
        }
        finally {
            this.logEvent('readFile', 'end', traceID, url);
        }
    }
    async readDirectory(url) {
        const traceID = performance.now();
        this.logEvent('readDir', 'start', traceID, url);
        try {
            checkCapabilityOrThrow(this.fs, this.capabilities, FSCapabilityFlags.ReadDir, 'readDirectory', url);
            return (await this.fs.readDirectory(url)).map((e) => new CVfsDirEntry(e));
        }
        catch (e) {
            this.logEvent('readDir', 'error', traceID, url, e instanceof Error ? e.message : '');
            throw wrapError(e);
        }
        finally {
            this.logEvent('readDir', 'end', traceID, url);
        }
    }
    async writeFile(file) {
        const traceID = performance.now();
        const url = file.url;
        this.logEvent('writeFile', 'start', traceID, url);
        try {
            checkCapabilityOrThrow(this.fs, this.capabilities, FSCapabilityFlags.Write, 'writeFile', file.url);
            return await this.fs.writeFile(file);
        }
        catch (e) {
            this.logEvent('writeFile', 'error', traceID, url, e instanceof Error ? e.message : '');
            throw wrapError(e);
        }
        finally {
            this.logEvent('writeFile', 'end', traceID, url);
        }
    }
    static disposeOf(fs) {
        fs instanceof WrappedProviderFs && fs.fs?.dispose();
    }
}
function checkCapabilityOrThrow(fs, capabilities, flag, name, url) {
    if (!(capabilities & flag)) {
        throw new VFSErrorUnsupportedRequest(name, url);
    }
}
class CFileType {
    fileType;
    constructor(fileType) {
        this.fileType = fileType;
    }
    isFile() {
        return this.fileType === FileType.File;
    }
    isDirectory() {
        return this.fileType === FileType.Directory;
    }
    isUnknown() {
        return !this.fileType;
    }
    isSymbolicLink() {
        return !!(this.fileType & FileType.SymbolicLink);
    }
}
class CVfsStat extends CFileType {
    stat;
    constructor(stat) {
        super(stat.fileType || FileType.Unknown);
        this.stat = stat;
    }
    get size() {
        return this.stat.size;
    }
    get mtimeMs() {
        return this.stat.mtimeMs;
    }
    get eTag() {
        return this.stat.eTag;
    }
}
export class CVfsDirEntry extends CFileType {
    entry;
    _url;
    constructor(entry) {
        super(entry.fileType);
        this.entry = entry;
    }
    get name() {
        return this.entry.name;
    }
    get dir() {
        return this.entry.dir;
    }
    get url() {
        if (this._url)
            return this._url;
        this._url = new URL(this.entry.name, this.entry.dir);
        return this._url;
    }
    toJSON() {
        return {
            name: this.name,
            dir: this.dir,
            fileType: this.fileType,
        };
    }
}
export function chopUrl(url) {
    if (!url)
        return '';
    const href = url.href;
    const parts = href.split('/');
    const n = parts.indexOf('node_modules');
    if (n > 0) {
        const tail = parts.slice(Math.max(parts.length - 3, n + 1));
        return parts.slice(0, n + 1).join('/') + '/…/' + tail.join('/');
    }
    return href;
}
export function rPad(str, len, ch = ' ') {
    return str.padEnd(len, ch);
}
